### PR TITLE
Use Windows hooks to block mouse input

### PR DESCRIPTION
**Pull request by ichisadashioko**
From the [Hooks Overview](https://docs.microsoft.com/en-us/windows/win32/winmsg/about-hooks), `hook` is capable of detecting mouse input message and intercepting them from reaching the "applications".

### CODE REVIEW

**Comment by ichisadashioko** [comment]
I wonder if the intercepting is per applications or system-wide? @alvinhochun

**Comment by ichisadashioko** [comment]
Someone mentioned that _global hook_ must be implemeted in a separated DLL and loaded at runtime. That sounds like a real pain as I don't know where the said DLL would be at runtime.

**Comment by alvinhochun** [comment]
You want to use the *low-level* mouse hook. The low-level mouse and keyboard hooks are special in that they run in the thread that installed the hook (with the help of the message queue) and are not injected into other processes, so there is no need to use a separate DLL at all. (see remarks in https://docs.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms644986(v=vs.85))

File: touchpad/main.c [hunk 1]
```diff
@@ -878,44 +903,22 @@ int CALLBACK wWinMain(_In_ HINSTANCE hInstance, _In_opt_ HINSTANCE hPrevInstance
...
-      // if (block_input_retval == 0)
-      //{
-      //  mGetLastError();
-      //}
-
-      ClipCursor(NULL);
+      if (llMouseHookProc != NULL)
+      {
+        // TODO how to unhook?
+        UnhookWindowsHookEx(hHook);
```
Comment (ichisadashioko) [inline touchpad/main.c]: @alvinhochun I think I got something wrong with the hooking flow. I was able to block mouse input system-wide but I can't unhook it when done.

Comment (ichisadashioko) [reply touchpad/main.c]: What is that `HHOOK` parameter that we need to pass into `UnhookWindowsHookEx`? Where do we get it?

Comment (alvinhochun) [reply touchpad/main.c]: You are confusing the two of `HOOKPROC` and `HHOOK`. `HHOOK` is the value returned from `SetWindowsHookEx` and `HOOKPROC` is the function pointer to your hook function. They are both defined as `void *` which is why your code typechecks, but it's definitely not correct.

Comment (ichisadashioko) [reply touchpad/main.c]: I got it.

**Pull request merged**